DOC: Adding tutorial for linear_sum_assignment algorithms and few generic problem statements

tapaswenipathak opened the pull request:
<!-- 
Thanks for contributing a pull request! Please ensure that
your PR satisfies the checklist before submitting:
http://scipy.github.io/devdocs/dev/contributor/development_workflow.html#checklist-before-submitting-a-pr

Also, please name and describe your PR as you would write a
commit message:
https://docs.scipy.org/doc/numpy/dev/gitwash/development_workflow.html#writing-the-commit-message

Note that we are a team of volunteers; we appreciate your
patience during the review process.

Again, thanks for contributing!
-->

#### Reference issue
Ref https://github.com/scipy/scipy/issues/10358.

#### What does this implement/fix?
Adding tutorials on few LSAP algorithms and generic problem statements.mostly using matrices http://www.assignmentproblems.com/doc/LSAPIntroduction.pdf.

#### Additional information
None.

tylerjereddy commented on doc/source/tutorial/optimize.rst:
```diff
@@ -1342,10 +1342,44 @@ residual is expensive to compute, good preconditioning can be crucial
+>>> from scipy.optimize import linear_sum_assignment
+>>> row_ind, col_ind = linear_sum_assignment(cost)
+>>> col_ind
+>>> array([1, 0, 2])
```
looks like refguide wants `np.array` here

tylerjereddy commented on doc/source/tutorial/optimize.rst:
```diff
@@ -1342,10 +1342,44 @@ residual is expensive to compute, good preconditioning can be crucial
+>>> col_ind
+>>> array([1, 0, 2])
+>>> cost[row_ind, col_ind].sum()
+>>> 5
```
If `5` is the expected value, we'll want to remove the starting `>>>` prompt; refguide is complaining about that in the CI

fuglede commented on doc/source/tutorial/optimize.rst:
```diff
@@ -1342,10 +1342,45 @@ residual is expensive to compute, good preconditioning can be crucial
+.. math::
+    \\min \\sum_i \\sum_j C_{i,j} X_{i,j}
+
+s.t. each row is assignment to at most one column, and each column to at most one row.
```
Passing by with a quick remark here: This addition duplicates the [docstring for the method as of its 1.3.0 version](https://github.com/scipy/scipy/blob/v1.3.0/scipy/optimize/_hungarian.py#L13-L118). Maybe that's okay? The docstring is accurate enough, but seems like it would be annoying to have to maintain the documentation in two places.

And note that by being based on the old version, it includes a [mistake that was fixed recently](https://github.com/scipy/scipy/pull/10344).

fuglede commented on doc/source/tutorial/optimize.rst:
```diff
@@ -1342,10 +1342,45 @@ residual is expensive to compute, good preconditioning can be crucial
+>>> from scipy.optimize import linear_sum_assignment
+>>> row_ind, col_ind = linear_sum_assignment(cost)
+>>> col_ind
+>>> np.array([1, 0, 2])
```
[Via Travis](https://travis-ci.org/scipy/scipy/jobs/564344199#L1395): There should be no `>>> ` here (you can imagine that you're writing the code in the Python REPL; there outputs are given on blank lines. As such, I would also imagine it should in fact be `array([1, 0, 2], dtype=int64)` here.

fuglede commented on doc/source/tutorial/optimize.rst:
```diff
@@ -1342,10 +1342,45 @@ residual is expensive to compute, good preconditioning can be crucial
+>>> col_ind
+>>> np.array([1, 0, 2])
+>>> cost[row_ind, col_ind].sum()
+
```
[Via Travis](https://travis-ci.org/scipy/scipy/jobs/564344199#L1410): The `5` should be moved one line up; outputs appear immediate below the inputs.

fuglede commented on doc/source/tutorial/optimize.rst:
```diff
@@ -1342,10 +1342,45 @@ residual is expensive to compute, good preconditioning can be crucial
+
+This function can also solve a generalization of the classic assignment problem where the cost matrix is rectangular. If it has more rows than columns, then not every row needs to be assigned to a column, and vice versa.
+
+>>> import scipy.optimize
```
This `import` is unused as you import the specific method a few lines below.

fuglede commented on doc/source/tutorial/optimize.rst:
```diff
@@ -1342,10 +1342,45 @@ residual is expensive to compute, good preconditioning can be crucial
+
+5
+
+Now we will use Hungarian method for solving a an assignment problem using
```
[Typo] "a an" should be "an". The PR is marked as a draft, so I guess there's also going to be something following this statement here?

fuglede commented on doc/source/tutorial/optimize.rst:
```diff
@@ -1342,10 +1342,45 @@ residual is expensive to compute, good preconditioning can be crucial
 lot more depth to this topic than is shown here.
 
+
+Linear Sum Assignment
```
This chapter should be in a section of its own; right now it's a subsection of "Root finding", but the two problems are not very related. This also means that the references end up looking like references for the linear sum assignment problem, when they're really not.

tapaswenipathak replied:
Sure.

tapaswenipathak commented:
Hi @fuglede: I created a draft PR, mind if I request review after I finish the tutorials. I have addressed few review comments you posted.

tapaswenipathak commented:
Hi @fuglede: I created a draft PR, mind if I request review after I finish the tutorials. I have addressed the review comments 02b39cdbc0050c2b4a275d540a0c74de96897cca.

mdhaber commented:
Hi @tapaswenipathak! Thanks for your interest in contributing to SciPy. 

As @fuglede mentioned, this is currently a copy of the `linear_sum_assignment` API reference. Tutorials serve a different purpose than the API reference: they are a gentle, narrative introduction to using a SciPy feature whereas the API reference concisely but comprehensively documents the interface.  For example, please compare the `scipy.optimize.linprog` [API reference](http://scipy.github.io/devdocs/generated/scipy.optimize.linprog.html#scipy.optimize.linprog) vs the new [tutorial](http://scipy.github.io/devdocs/tutorial/optimize.html#linear-programming-linprog).

Since this PR now has merge conflicts, I will probably close it in about a week unless you would like to finish this. Please let me know.

@AtsushiSakai , if @tapaswenipathak is not interested in completing this, would you like to start one?

AtsushiSakai commented:
@mdhaber Yes. If this PR is closed, I will try it.

mdhaber commented:
Thanks @AtsushiSakai. That would be great.
